File: 0014_Longest_Common_Prefix/Longest_Common_Prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty()){      // Checking if even there is a string or not
            return "";
        }

        string prefix = strs[0];        // Saving first element in string type variable

        for(int i = 0; i < strs.size(); i++){       // Looping through each element

            while (strs[i].compare(0, prefix.size(), prefix) != 0 && !prefix.empty()){      // !prefix.empty() is not needed. the previous condition is sufficient but i still added it 😁. What it does is compare prefix with string until it matches
                prefix.pop_back();          // If not matches then removes the last element of the prefix. No need to re-assign. works directly at the memory.
                
                if(prefix.empty())              // Returns in boolean
                    return "";      // Eventually if the string goes empty then return ""
            }
        }
        return prefix;          // if something is left then that is returned
    }
};

File: 0014_Longest_Common_Prefix/Longest_Common_Prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            # If there is nothing from the start then why should i try, just return the empty string
            return ""

        # Taking the first string of the list of string as starting point.
        prefix = strs[0]
        # later we will compare the rest of the strings

        # Here is not indeces. it is representing every elements/strings of the strs string from index 1
        for s in strs[1:]:

            # This loop will continue until match is found
            while prefix and not s.startswith(prefix):
                # Every time prefix does not matches it reduces one element of the string from the end
                prefix = prefix[:-1]

            if not prefix:
                # If  nothing is left after loop then retun this. else we will compare rest elements with leftover of prefix
                return ""

        return prefix  # If something is left, then returning it
